cosmos/lcd.go

```diff
@@ -0,0 +1,225 @@
+func (lcd *LCD) Post(path string, req interface{}, ptr interface{}) error {
+	cosResp := rest.ResponseWithHeight{}
+	if err := lcd.PostBare(path, req, &cosResp); err != nil {
+		return err
+	}
+	if len(cosResp.Result) > 0 {
+		if err := lcd.cdc.UnmarshalJSON(cosResp.Result, ptr); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// PostBare executes a post request on the LCD.
+// There is no expectation on the type of response.
+func (lcd *LCD) PostBare(path string, req interface{}, ptr interface{}) error {
+	reqBody, err := lcd.cdc.MarshalJSON(req)
+	if err != nil {
+		return err
+	}
+	resp, err := http.Post(lcd.endpoint+path, "application/json", bytes.NewReader(reqBody))
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return fmt.Errorf("request status code is not 2XX, got %d", resp.StatusCode)
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	if err := lcd.cdc.UnmarshalJSON(body, ptr); err != nil {
+		return err
+	}
+	return nil
+}
+
+//BroadcastMsg sign and broadcast a transaction from a message.
+func (lcd *LCD) BroadcastMsg(msg sdk.Msg) ([]byte, error) {
+	return lcd.BroadcastMsgs([]sdk.Msg{msg})
+}
+
+//BroadcastMsgs sign and broadcast a transaction from multiple messages.
+func (lcd *LCD) BroadcastMsgs(msgs []sdk.Msg) ([]byte, error) {
+	tx, err := lcd.createAndSignTx(msgs)
+	if err != nil {
+		return nil, err
+	}
+	req := authrest.BroadcastReq{
+		Tx:   tx,
+		Mode: "block", // TODO: should be changed to "sync" and wait for the tx event
+	}
+	var res sdk.TxResponse
+	if err := lcd.PostBare("txs", req, &res); err != nil {
+		return nil, err
+	}
+	if abci.CodeTypeOK != res.Code {
+		return nil, fmt.Errorf("transaction returned with invalid code %d: %s", res.Code, res)
+	}
+	result, err := hex.DecodeString(res.Data)
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+func (lcd *LCD) getAccount() (*auth.BaseAccount, error) {
+	lcd.getAccountMutex.Lock()
+	defer lcd.getAccountMutex.Unlock()
+	if lcd.acc == nil {
+		accKb, err := lcd.kb.Get(lcd.accName)
+		if err != nil {
+			return nil, err
+		}
+		lcd.acc = auth.NewBaseAccount(accKb.GetAddress(), nil, accKb.GetPubKey(), 0, 0)
+	}
+	localSeq := lcd.acc.GetSequence()
+	if err := lcd.Get("auth/accounts/"+lcd.acc.GetAddress().String(), &lcd.acc); err != nil {
+		return nil, err
+	}
+	// replace seq if sup
+	if localSeq > lcd.acc.GetSequence() {
+		lcd.acc.SetSequence(localSeq)
+	}
+	return lcd.acc, nil
+}
+
+func (lcd *LCD) createAndSignTx(msgs []sdk.Msg) (authtypes.StdTx, error) {
+	// retrieve account
+	accR, err := lcd.getAccount()
+	if err != nil {
+		return authtypes.StdTx{}, err
+	}
+	sequence := accR.GetSequence()
+	accR.SetSequence(accR.GetSequence() + 1)
+
+	// Create TxBuilder
+	txBuilder := authtypes.NewTxBuilder(
+		authutils.GetTxEncoder(lcd.cdc),
+		accR.GetAccountNumber(),
+		sequence,
+		flags.DefaultGasLimit*10,
```

Comment:
why `*10`?

Reply:
because of the transaction that contains 10 executions. the default was not enough

Reply:
it will be removed when estimate gas is implemented

Reply:
if we have this generic client we should then pass this as an argument to be able to customize it.
Can have multiple ways:
- as a parameter of this function and so all the broadcast functions
- as an `option` function on the LCD that will write it int the attributes of the client `WithGazLimit()` that will be default `flags.DefaultGasLimit`
- as a parameter of the constructor like it's done for the `minGasPrices`

I'm fine with any of these solutions but hardcoding a `*10` is a bit sketchy for a generic client

Reply:
I add a parameter `gasLimit` to the LCD